plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs

```diff
@@ -129,33 +129,89 @@ instance Pretty ExCPU where
+
+This is pretty reasonable: a polymorphic builtin *is* allowed to check if the AST that it got as an
+argument is a constant or not, and if it happens to be a constant, the builtin *is* allowed to use
+it whatever way it wishes (see Note [Builtins and Plutus type checking] for details). Hence a
+builtin may in fact do something ad hoc for constants and we need to account for this possibility in
+the costing machinery.
+
+But if the given AST is not a constant, the builtin can't do anything else with it, hence we simply
+return 1, meaning "the costing function can't use this 'ExMemory' in any non-vacuous way".
+
+See 'HasMeaningIn' for a full list of constraints determining what a builtin can do with values.
+
+However for all types of values, except the one used by the production evaluator, we implement
+'ExMemoryUsage' as a call to 'error'. Not because other evaluators don't compute costs during
+evaluation -- the CK machine for example does in fact compute them (because we share the same
+builtins machinery between all the evaluators and we want it to be efficient on the production path,
+hence it's easier to optimize it for all evaluators than just for the single production evaluator).
+And not because the resulting 'ExBudget' is not forced by an evaluator that doesn't care about
+costing -- it still gets forced (for the same reason).
+
+The actual reason why we call 'error' is because at the moment no builtin is supposed to have a
+costing function that actually computes the 'ExMemory' of the given AST. Currently, if the builtin
+takes an 'Opaque', it's not supposed to actually look inside of it (unlike with 'SomeConstant') and
+hence the costing function is supposed to ignore that argument. It is possible that we'll eventually
+decide to add such a builtin, so the current approach of throwing an 'error' is a precaution
+ensuring that we won't add any weirdness by accident.
+
+We don't call 'error' on the production path, because we don't want this risk in there. A failing
+test is fine, a failing reasonable transaction is not and we don't want to risk it, even if it seems
+very unlikely that such a failure could slip in.
+
+The way we ignore arguments in costing functions is by computing the 'ExMemory' of each of those
+arguments lazily. I.e. a call to 'memoryUsage' can only be forced within a costing function and
+never outside of one. We have to do this regardless of all the reasoning above: if we compute
+the 'ExMemory' of, say, a list strictly, then a builtin prepending an element to a list will
+have the complexity of O(length_of_the_list) (because computing the 'ExMemory' of a list requires
+traversing the list), while we of course want it to be O(1).
+-}
+
 class ExMemoryUsage a where
     memoryUsage :: a -> ExMemory -- ^ How much memory does 'a' use?
 
 instance (ExMemoryUsage a, ExMemoryUsage b) => ExMemoryUsage (a, b) where
     memoryUsage (a, b) = 1 <> memoryUsage a <> memoryUsage b
+    {-# INLINE memoryUsage #-}
```

Comment:
> Adding those pragmas alone causes a slowdown: https://github.com/input-output-hk/plutus/pull/4765.

I'm not sure what you mean by "alone". What's the difference between this and #4675?  Is something else happening somewhere that stops the slowdown?

Reply:
In the other PR calls to `memoryUsage` don't get inlined for specific types, because in the current version of the builtin application machinery we only check memory usages of the whole `CekValue` provided as an argument. Meaning GHC sees the argument as a `CekValue` with an arbitrarily typed constant inside and thus GHC can't inline much.

With this PR we only feed properly unlifted values to costing functions, so that GHC knows the type of each of those values and can pick an appropriate `ExMemoryUsage` instance at runtime and inline the call to `memoryUsage` away.

So I don't know what causes the slowdown, but I know what causes the speedup (it was the whole point of this refactoring).